Googleplus Image Support

anotherdirtbag opened the pull request:
## Description:
Adding support for getting a googleplus thumbnail image from their email. This is very similar to the existing gravitar support but can fail if a non-google account or invalid email is used. it currently prioritizes gravitar over googleplus over picture

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3999

## Example entry for `configuration.yaml` (if applicable):
```yaml
robbiesiphone:
  name: robbiesiphone
  googleplus: robbie@gmail.com
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

anotherdirtbag committed:
Adding support for getting a googleplus thumbnail image from their email. This is very similar to the existing gravitar support but can fail if a non-google account or invalid email is used. it currently prioritizes gravitar over googleplus over picture

homeassistant commented:
Hi @anotherdirtbag, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#10615).

Once you do that we will be able to review and accept this pull request.

Thanks!

anotherdirtbag committed:
googleplus thumbnail: fixed some warnings

anotherdirtbag commented:
I don't see a REQUIREMENTS variable in either file I edited. 
I added urllib and re to \homeassistant\components\device_tracker\__init__.py
I also added re to \tests\components\device_tracker\test_init.py

anotherdirtbag committed:
googleplus thumbnail: fixing warnings

anotherdirtbag committed:
googleplus thumbnail: fixed tests

anotherdirtbag committed:
googleplus thumbnail: changed imports slightly

anotherdirtbag commented:
to answer my own question, I don't think i need to edit the REQUIREMENTS since the libraries i'm importing are standard python 3 stuff.

anotherdirtbag committed:
googleplus thumbnail: fixing warnings

anotherdirtbag committed:
googleplus thumbnail: config_picture is actually set now

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

anotherdirtbag committed:
googleplus thumbnail: moved annoying code into try catch to hopefully avoid the warning

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

anotherdirtbag commented:
3 commits that modify the code plus 10 commits and counting to jump through the formatting hoops for 30 lines of code plus a unit test.
I need a try catch to catch the non str types that were being passed into urllib.parse.quote even though the input is forced to be str as far as I understand. None and bool are the ones I'm aware of so far. Now there's a warning that bare exceptions are not allowed.

homeassistant/components/device_tracker/__init__.py
W:771, 4: No exception type(s) specified (bare-except)

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

Danielhiversen commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -747,3 +756,32 @@ def get_gravatar_for_email(email: str):
+        url += urllib.parse.quote(email) + r'?alt=json'
+        jsontext = urllib.request.urlopen(url).read().decode('utf-8')
+    except TypeError:
+        print(email)
```
Don't use print. You can use LOGGER.error

Spartan-II-117 replied:
@danielhiverson does this satisfy your change request?

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

anotherdirtbag committed:
googleplus thumbnail: dodging a type warning

anotherdirtbag committed:
googleplus thumbnail: still fixing warnings

anotherdirtbag commented:
the travis lint build timed out after 20 minutes. I'll let you guys take it from here

Danielhiversen commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -747,3 +756,32 @@ def get_gravatar_for_email(email: str):
+
+def get_googlepluspthumb_from_email(email: str):
+    """Return a thumbnail image of a user's google plus profile image."""
+    import urllib.request
```
This should be changed to aiohttp
See get_vendor_for_mac() for an example

anotherdirtbag commented:
I'm not figuring out how to get aiohttp to work in my testing. something about making a clientsession in a separate thread. From what you said, it looks like i can just replace
`jsontext = urllib.request.urlopen(url).read().decode('utf-8')`
with
`websession = async_get_clientsession(self.hass)
        with async_timeout.timeout(5, loop=self.hass.loop):
            resp = yield from websession.get(url)
            if resp.status == 200:
                jsontext = yield from resp.text()`
but I'm not able to test it to see if that actually works or not. I'll also have to add cases for asyncio.TimeoutError and aiohttp.ClientError.

anotherdirtbag commented:
actually, I could test it by temporarily adding some print commands to the unit test and looking at the travis build output. that should work.

anotherdirtbag committed:
using aiohttp instead of urllib.request

anotherdirtbag committed:
fixing tabs and spaces to see debug output

anotherdirtbag committed:
minor fix to see debug output

anotherdirtbag committed:
fixing self types

anotherdirtbag commented on homeassistant/components/device_tracker/__init__.py:
```diff
@@ -739,11 +748,50 @@ def update_config(path: str, dev_id: str, device: Device):
+            resp = yield from websession.get(url)
+            print('resp.status=' + resp.status)
+            jsontext = yield from resp.text()
+            debug_jsontext=yield from resp.json(encoding = 'utf-8')
```
lol. pick one. missing spaces, or unexpected spaces. it shouldn't be both

Danielhiversen replied:
Try
debug_jsontext = yield from resp.json(encoding='utf-8')

anotherdirtbag replied:
thanks. with two assignments it makes sense now.

anotherdirtbag committed:
fixing types

anotherdirtbag committed:
fixing types

anotherdirtbag committed:
googleplus thumbnail: fixing whitespace warnings

anotherdirtbag committed:
googleplus thumbnail: fixing whitespace warnings

anotherdirtbag committed:
fixing types

anotherdirtbag committed:
googleplus thumbnail: fixing whitespace warnings

anotherdirtbag committed:
converting generator type to hopefully a string type with ''.join(list(generator))

anotherdirtbag committed:
googleplus thumbnail: fixing whitespace warnings

anotherdirtbag commented:
If this doesn't work, I'm not sure what will.
To fix the current type error
`TypeError: __init__() should return None, not 'generator'`
I tried to convert the generator to a string type using
https://stackoverflow.com/questions/14447081/python-generator-objects-and-join

anotherdirtbag committed:
googleplus thumbnail: a different method for converting generator type to a string

anotherdirtbag commented:
after looking into it more, the asyncio stuff is beyond me. 
I think, no matter what, I'm not allowed to convert the results of 'yield from' to a normal string and instead have to call it from one of the asyncio coroutines except that __init__ where all of the device config/initialization happens is not an async routine.
someone else can look at this, or we could roll back to 318849753c70c8dade5c22c32f8964bc1923d151 which used urllib and everything worked.

anotherdirtbag committed:
googleplus thumbnail: removed debugging/test code. get_googlepluspthumb_from_email still doesn't work

anotherdirtbag committed:
googleplus thumbnail: trying something else

Spartan-II-117 commented:
@Danielhiversen  any guidance?

balloob commented:
Doing a regex on a page to get a picture is not something that should be added to the core of Home Assistant. Gravatar support should also probably be removed. It's a mess.

We should not merge this and the PR has gone stale. Closing it.
